ssl/ssl_lib.c

```diff
@@ -1009,7 +1017,7 @@ void SSL_free(SSL *s)
     if (s == NULL)
         return;
 
-    i = CRYPTO_add(&s->references, -1, CRYPTO_LOCK_SSL);
+    i = CRYPTO_atomic_add(&s->references, -1, &i, s->lock);
```

Comment:
This line is the cause of your memory leaks. Don't assign the result of CRYPTO_atomic_add to `i`.